### DISCUSSION TITLE
Why no trade after the signal received?

### DISCUSSION THREAD

**Discussion by AetherTL** [kernc/backtesting.py#486] [General]
![截屏2021-09-30 上午10 17 05](https://user-images.githubusercontent.com/23423871/135375065-9a751eec-5a42-4bd7-afba-72391b28b05e.png)

The entry size plot shows a short signal was received at around March 1987, but no trade has been made, I don't know why

I was using the HSI index as the dataset, and the strategy is simple, if the Z-value < -threshold the short the index, if >threshold then long it, it works at Oct 87 but failed at March, I am confusing now

**Comment by AetherTL** [comment]
![截屏2021-09-30 上午10 51 47](https://user-images.githubusercontent.com/23423871/135378074-b9a6a1d9-26d4-4a0c-af72-4a3ea28bdbb8.png)
It seems the first trade signal always been ignored, but why? Is this designed or a bug? How can I fix this?

**Reply by kernc** [reply]
This is [`SignalStrategy` code](https://github.com/kernc/backtesting.py/blob/7671f044d554249b608b3241dd7cca37ec5a35d1/backtesting/lib.py#L368-L437). Welcome to investigate. :sweat_smile: 

> It seems the first trade signal always been ignored, but why?

`Z.shift()` produces one leading `NaN` value, but `NaN != 0`, hence +1. Can you confirm?

**Comment by AetherTL** [comment]
thanks for your reply

The signal list is perfectly matched so I don’t think Z.shift() matters here, also Z.shift only =nan at first date right? In my example the first signal was received at 43rd bar.

**Reply by kernc** [reply]
Referring to:

> It seems the first trade signal always been ignored, but why?

You do `print(sum(signal.head(100) != 0))` and get 40, whereas there are only 39 trades. The extra one is the 0-indexed NaN.

**Comment by AetherTL** [comment]
![截屏2021-10-04 上午9 51 17](https://user-images.githubusercontent.com/23423871/135782639-32327c26-96c0-4808-9263-df5513be7f7f.png)

There is no Nan in the signal :(, 6 signals and 5 trades

You can see my full code in colab:

https://colab.research.google.com/drive/1l4Ki7tcdlhjEqfcuhdETdxbupIwyB6KS?usp=sharing

**Reply by kernc** [reply]
Opened this as issue https://github.com/kernc/backtesting.py/issues/495. Welcome to subscribe and/or investigate further. Note that [`SignalStrategy`](https://kernc.github.io/backtesting.py/doc/backtesting/lib.html#backtesting.lib.SignalStrategy) doesn't close trades unless `exit_portion=` is set—rather it just uses [`Strategy.buy`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.buy) and `Strategy.sell` ...
https://github.com/kernc/backtesting.py/blob/7671f044d554249b608b3241dd7cca37ec5a35d1/backtesting/lib.py#L423-L437
So this may play a role. :thinking:
